### ISSUE TITLE
TypeVarTuple support for `*args` doesn't handle unpacked tuple with suffix

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4119]
This is a follow-on to https://github.com/microsoft/pyright/issues/4117.

```python
from typing_extensions import TypeVarTuple, Unpack

Ts = TypeVarTuple("Ts")

def process(*args: Unpack[tuple[Unpack[Ts], int]]):
    ...

def process2(*args: Unpack[tuple[Unpack[Ts]]]):
    ...

process(*(1, 2, 3))  # error: Argument missing for parameter "args[1]"
process2(*(1, 2, 3))  # no error
```

**Comment by erictraut** [comment]
This is also a problem for prefixes. There is no error in the case of a prefix, but the type matched to the TypeVarTuple is incorrect.

```python
from typing_extensions import TypeVarTuple, Unpack

Ts = TypeVarTuple("Ts")

def process_simple(*args: Unpack[tuple[Unpack[Ts]]]) -> tuple[*Ts]:
    ...

def process_prefix(*args: Unpack[tuple[int, Unpack[Ts]]]) -> tuple[*Ts]:
    ...

def process_suffix(*args: Unpack[tuple[Unpack[Ts], float]]) -> tuple[*Ts]:
    ...

def func(x: int, y: str, z: float):
    v1 = process_simple(*(x, y, z))
    reveal_type(v1, expected_text="tuple[int, str, float]")

    v2 = process_prefix(*(x, y, z))
    reveal_type(v2, expected_text="tuple[str, float]") # false positive

    v3 = process_suffix(*(x, y, z)) # false positive
    reveal_type(v3, expected_text="tuple[int, str]") # false positive
```

**Comment by mehdigmira** [comment]
Have a new issue that i believe is closely related to this, and it can be of any use while fixing this:
```
from typing import Generic

from typing_extensions import TypeVarTuple, Unpack

Ts = TypeVarTuple("Ts")


class A(Generic[Unpack[Ts]]):
    def f(self, *args: Unpack[tuple[Unpack[Ts], str]], other: str):
        ...


a = A[int, int]()

a.f(1, 2, "s", "s")  # no error here while `other` is kw only
```

**Comment by erictraut** [comment]
@mehdigmira, this is a separate bug, so please create a new issue to track it. Thanks for finding and reporting these. The code related to argument/parameter matching is incredibly complex in Python type checkers (with positional parameters, keyword parameters, *args, **kwargs, arg unpacking, default argument values, etc.). PEP 647 introduced more than doubled that complexity by introducing TypeVarTuple and unpacked tuples with prefixes and suffixes. I'm not surprised that you're finding a bunch of edge cases. When PEP 647 was in draft form, I warned the authors that they were going to create a lot of work for all of the Python type checker maintainers, and I was right about that prediction.

**Comment by erictraut** [comment]
This is included in pyright 1.1.279, which I just published. It will also be included in a future version of pylance.
